rfd/0074-kube-secret-storage.md

```diff
@@ -0,0 +1,603 @@
+      storageClassName: {{ .Values.storage.storageClassName }}
+      resources:
+        requests:
+          storage: {{ .Values.storage.requests }}
++{{- end }}
+{{- end }}
+```
+
+#### File *values.yaml*
+
+Change storage comments
+
+### File *post-upgrade-job.yaml*
+```diff
++{{- $deployment := lookup "v1" "Deployment" .Release.Namespace .Release.Name -}}
++{{- if ( $deployment ) }}
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: "{{ .Release.Name }}"
+  labels:
+    app.kubernetes.io/managed-by: {{ .Release.Service | quote }}
+    app.kubernetes.io/instance: {{ .Release.Name | quote }}
+    app.kubernetes.io/version: {{ .Chart.AppVersion }}
+    helm.sh/chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+  annotations:
+    "helm.sh/hook": post-upgrade
+    "helm.sh/hook-weight": "-5"
+    "helm.sh/hook-delete-policy": hook-succeeded
+spec:
+  template:
+    metadata:
+      name: "{{ .Release.Name }}"
+      labels:
+        app.kubernetes.io/managed-by: {{ .Release.Service | quote }}
+        app.kubernetes.io/instance: {{ .Release.Name | quote }}
+        helm.sh/chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: post-upgrade-job
+        image: "alpine:3.3"
+        command: ["/bin/bash","-c", "kubectl rollout status --watch --timeout=600s statefulset/{{ .Release.Name }} && kubectl delete deployment/{ .Release.Name } -n { .Release.Namespace }"]
++{{- end}}
+```
+
+
+## UX
+
+If the agent detects that it is running inside Kubernetes, it will enable, by default, the Kube secret storage instead of local storage. This means that it will not require any additional configuration or operation by the end-user.
+
+To install the agent on Kubernetes using Helm, the command remains the same as before, as one can see below.
+
+```bash
+$ helm install/upgrade teleport-agent  teleport/teleport-kube-agent \
+  --create-namespace \
+  --namespace teleport-agent \
+  --set roles=kube \
+  --set proxyAddr=${PROXY_ENDPOINT?} \
+  --set authToken=${JOIN_TOKEN?} \
+  --set kubeClusterName=${KUBERNETES_CLUSTER_NAME?}
+```
+
+## Security
+
+The change with the greatest impact is that the agent's service account has read and write permissions for any existing secret in the namespace where it was installed. If the SA token is stolen or a hijack happens in the Teleport agent, the attacker can use it to read any secrets in that namespace. We recommend that Teleport is installed in a separate namespace to minimize this risk.
+
+This issue happens because, currently, Kubernetes does not allow the specification of `resourceNames` in RBAC roles with [wildcards][wildcards]. If supported in the future, we can limit the read/list/update actions when handling secrets for the Teleport Service account. Otherwise, another possible solution might be to keep the identity stored in a single Secret instead of a secret per pod, allowing read and write operations only for that secret.
```

Comment:
Another solution should be studied to tackle this in case of (hopefully) k8s RBAC design changes. The feature theorized here was already discussed and declined in https://github.com/kubernetes/kubernetes/issues/56582 because considered to be too fragile:
>Even if RBAC was more expressive, the security boundary is still the namespace. [...] Letting user bob manage all namespaces that match `bob-*` sounds great until user `bob-tables` shows up.

and to be honest I can see the maintainers' point.